File: lynx/render/touch/touch_event.cpp
// Copyright 2017 The Lynx Authors. All rights reserved.

#include "runtime/base/lynx_value.h"
#include "render/touch/touch_event.h"

namespace lynx {

    TouchEvent::TouchEvent(TouchEventInfo& info) : JSEvent(info.touch_event_type()),
                                                   touch_event_type_(info.touch_event_type()),
                                                   motion_event_type_(info.motion_event_type()),
                                                   x_(info.GetX()),
                                                   y_(info.GetY()) {
        set_class_name("TouchEvent");
        RegisterAccessorCallback("changedTouches", &GetTouchesCallback, 0);
        RegisterAccessorCallback("touches", &GetTouchesCallback, 0);
    }

    TouchEvent::~TouchEvent() {

    }

    void TouchEvent::Recycle() {

    }

    TouchEvent* TouchEvent::Obtain(TouchEventInfo info) {
        TouchEvent* event = lynx_new TouchEvent(info);
        return event;
    }

    base::ScopedPtr<jscore::LynxValue> TouchEvent::GetTouchesCallback(jscore::LynxObjectTemplate* object) {
        jscore::LynxArray* touches = lynx_new jscore::LynxArray();
        jscore::LynxObject* axis = lynx_new jscore::LynxObject();

        float x = static_cast<TouchEvent*>(object)->x();
        float y = static_cast<TouchEvent*>(object)->y();
        jscore::LynxValue* x_value = jscore::LynxValue::MakeFloat(x);
        jscore::LynxValue* y_value = jscore::LynxValue::MakeFloat(y);
        axis->Set("clientX", x_value);
        axis->Set("clientY", y_value);
        axis->Set("screenX", x_value);
        axis->Set("screenY", y_value);

        touches->Push(axis);
        return jscore::LynxValue::MakeValueScoped(touches);
    }

}

File: Android/sdk/src/main/java/com/lynx/ui/event/TouchEvent.java
// Copyright 2017 The Lynx Authors. All rights reserved.
package com.lynx.ui.event;

import android.util.SparseArray;

import com.lynx.core.base.LynxArray;
import com.lynx.core.base.LynxEvent;
import com.lynx.core.base.LynxHolder;
import com.lynx.core.base.LynxMap;
import com.lynx.core.touch.TouchEventInfo;
import com.lynx.ui.LynxUI;
import com.lynx.utils.PixelUtil;

/* package */ class TouchEvent extends LynxEvent {

    private final static String KEY_CHANGED_TOUCHES = "changedTouches";
    private final static String KEY_TOUCHES = "touches";
    private final static String KEY_CLIENT_X = "clientX";
    private final static String KEY_CLIENT_Y = "clientY";
    private final static String KEY_SCREEN_X = "screenX";
    private final static String KEY_SCREEN_Y = "screenY";
    private final static String KEY_SRC_ELEMENT = "srcElement";

    public long timeStamp;

    public TouchEvent(TouchEventInfo info) {
        super(info.getType());
        init(info);
    }

    private void init(TouchEventInfo info) {
        timeStamp = info.getTimeStamp();
        initChangedTouches(info);
        initTouches(info);
        setTarget(((LynxUI) info.getCurTarget()).getRenderObjectImpl());
        setSrcElement(((LynxUI) info.getTarget()).getRenderObjectImpl());
    }

    private void setSrcElement(LynxHolder srcElement) {
        set(KEY_SRC_ELEMENT, srcElement);
    }

    private void initChangedTouches(TouchEventInfo info) {
        // Add extra property
        LynxArray changedTouches = new LynxArray();
        set(KEY_CHANGED_TOUCHES, changedTouches);
        SparseArray<TouchEventInfo> infoList = info.getChangedInfoList();
        for (int i = 0; i < infoList.size(); ++i) {
            info = infoList.valueAt(i);
            LynxMap axis = new LynxMap();
            setClientAxis(axis,
                    (float) PixelUtil.pxToLynxNumber(info.getX()),
                    (float) PixelUtil.pxToLynxNumber(info.getY()));
            setScreenAxis(axis,
                    (float) PixelUtil.pxToLynxNumber(info.getRawX()),
                    (float) PixelUtil.pxToLynxNumber(info.getRawY()));
            changedTouches.add(axis);
        }
    }

    private void initTouches(TouchEventInfo info) {
        // Add extra property
        LynxArray touches = new LynxArray();
        set(KEY_TOUCHES, touches);
        SparseArray<TouchEventInfo> infoList = info.getInfoListOnScreen();
        for (int i = 0; i < infoList.size(); ++i) {
            info = infoList.valueAt(i);
            LynxMap axis = new LynxMap();
            setClientAxis(axis,
                    (float) PixelUtil.pxToLynxNumber(info.getX()),
                    (float) PixelUtil.pxToLynxNumber(info.getY()));
            setScreenAxis(axis,
                    (float) PixelUtil.pxToLynxNumber(info.getRawX()),
                    (float) PixelUtil.pxToLynxNumber(info.getRawY()));
            touches.add(axis);
        }
    }

    private void setClientAxis(LynxMap axis, float x, float y) {
        axis.set(KEY_CLIENT_X, x);
        axis.set(KEY_CLIENT_Y, y);
    }

    private void setScreenAxis(LynxMap axis, float x, float y) {
        axis.set(KEY_SCREEN_X, x);
        axis.set(KEY_SCREEN_Y, y);
    }

    @Override
    public int getCoalescingKey() {
        return 1;
    }

    @Override
    public boolean canCoalesce() {
        switch (mType) {
            case TouchEventInfo.START:
            case TouchEventInfo.END:
            case TouchEventInfo.CANCEL:
                return false;
            case TouchEventInfo.MOVE:
                return true;
            default:
                throw new RuntimeException("The touch event can not be recognized");
        }
    }
}
